### ISSUE TITLE
failure to generate dynamic agents in Pedestrian Navigation (console mode) example on Linux

### ISSUE THREAD

**Issue by ABallisat** [FLAMEGPU/FLAMEGPU#228]
The Pedestrian Navigation (console mode) will compile and run on Linux but it does not generate any dynamic agents. This was found on 64-bit CentOS 7 using CUDA 7.5. The graphics card used was a P1000.

**Comment by ptheywood** [comment]
I can confirm that the Pedestrian models (both PedestrianLOD and PedestrianNavigation) do not create any pedestrian agents in console mode.

Pedestrians are created based on probabilistic values `EMMISION_RATE_EXIT1`, `EMMISION_RATE_EXIT2`, etc. 

These values are set to non zero values in [`src/visualisation/GlobalsController.cpp:initGlobalsController`](https://github.com/FLAMEGPU/FLAMEGPU/blob/master/examples/PedestrianNavigation/src/visualisation/GlobalsController.cpp#L94), which is called by [`src/visualisation/CustomVisualisation.cu:initVisualisation`](https://github.com/FLAMEGPU/FLAMEGPU/blob/a1a9c777806cd4e7024b8dbcde65d9ee454f35c2/examples/PedestrianLOD/src/visualisation/CustomVisualisation.cpp#L88).

This is only executed in visualisation mode, resulting in emission rates of `0.0` in console mode. 


There are several options for fixing this.
1. Using the upcoming 1.5.0 version of FLAME GPU, the initial value can be specified in the `<environment>` portion of the initial states file.
2. Using the upcoming 1.5.0 version of FLAME GPU, environment constants can be provided with a `defaultValue` in `XMLModelFile.XML`, which will be used if a value is not provided in the initial states file
3. Some of the behaviour contained within `GlobalsController` should be moved to an INIT function, which will be executed in both Console and Visualisation modes.

**Comment by ptheywood** [comment]
After speaking to @mondus the plan is to move the initialisation of environmental constants from the custom visualisation into the initial states XML file, which will  only work for FLAME GPU 1.5 and above, which will be released in the coming weeks. 

If you require a fix in the 1.4.x version of FLAME GPU, then add a new `FLAMEGPU_INIT_FUNC` to both `XMLModelFile.xml` and `functions.c`, which duplicates the initialisation of values from `visualisation/GlobalsController` to initialise emission rates. 

The following code is untested and not ideal due to duplication of behaviour, but *should* provide a quick fix. 

### XMLModeFile.xml
In the `<gpu:environment>`, after `</gpu:constants>` add

```
<gpu:initFunctions>
    <gpu:initFunction>
        <gpu:name>setInitialConstants</gpu:name>
    </gpu:initFunction>
</gpu:initFunctions>
```

### functions.c

In funnctions.c, add the following function

```
#define INITIAL_EMISSION_RATE_EXIT1 10.f
#define INITIAL_EMISSION_RATE_EXIT2 10.f
#define INITIAL_EMISSION_RATE_EXIT3 10.f
#define INITIAL_EMISSION_RATE_EXIT4 10.f
#define INITIAL_EMISSION_RATE_EXIT5 10.f
#define INITIAL_EMISSION_RATE_EXIT6 10.f
#define INITIAL_EMISSION_RATE_EXIT7 10.f
#define EMISSION_RATE_INCREMENT 0.01f

#define INITIAL_EXIT1_PROBABILITY 1
#define INITIAL_EXIT2_PROBABILITY 1
#define INITIAL_EXIT3_PROBABILITY 1
#define INITIAL_EXIT4_PROBABILITY 1
#define INITIAL_EXIT5_PROBABILITY 3
#define INITIAL_EXIT6_PROBABILITY 2
#define INITIAL_EXIT7_PROBABILITY 1

#define INITIAL_EXIT1_STATE 1
#define INITIAL_EXIT2_STATE 1
#define INITIAL_EXIT3_STATE 1
#define INITIAL_EXIT4_STATE 1
#define INITIAL_EXIT5_STATE 1
#define INITIAL_EXIT6_STATE 1
#define INITIAL_EXIT7_STATE 1

#define INITIAL_TIME_SCALER     0.0003
#define TIME_SCALER_INCREMENT   0.00001

#define INITIAL_STEER_WEIGHT        0.10
#define INITIAL_AVOID_WEIGHT        0.02
#define INITIAL_COLLISION_WEIGHT    0.50
#define INITIAL_GOAL_WEIGHT         0.20

#define STEER_WEIGHT_INCREMENT      0.001
#define AVOID_WEIGHT_INCREMENT      0.001
#define COLLISION_WEIGHT_INCREMENT  0.001
#define GOAL_WEIGHT_INCREMENT       0.001

#define EXIT1_CELL_COUNT 16
#define EXIT2_CELL_COUNT 26
#define EXIT3_CELL_COUNT 31
#define EXIT4_CELL_COUNT 24
#define EXIT5_CELL_COUNT 20
#define EXIT6_CELL_COUNT 66
#define EXIT7_CELL_COUNT 120


__FLAME_GPU_INIT_FUNC__ void setInitialConstants(){
    float emmisionRateExit1 = INITIAL_EMISSION_RATE_EXIT1;
    float emmisionRateExit2 = INITIAL_EMISSION_RATE_EXIT2;
    float emmisionRateExit3 = INITIAL_EMISSION_RATE_EXIT3;
    float emmisionRateExit4 = INITIAL_EMISSION_RATE_EXIT4;
    float emmisionRateExit5 = INITIAL_EMISSION_RATE_EXIT5;
    float emmisionRateExit6 = INITIAL_EMISSION_RATE_EXIT6;
    float emmisionRateExit7 = INITIAL_EMISSION_RATE_EXIT7;

    int exitProbability1 = INITIAL_EXIT1_PROBABILITY;
    int exitProbability2 = INITIAL_EXIT2_PROBABILITY;
    int exitProbability3 = INITIAL_EXIT3_PROBABILITY;
    int exitProbability4 = INITIAL_EXIT4_PROBABILITY;
    int exitProbability5 = INITIAL_EXIT5_PROBABILITY;
    int exitProbability6 = INITIAL_EXIT6_PROBABILITY;
    int exitProbability7 = INITIAL_EXIT7_PROBABILITY;

    int exitState1 = INITIAL_EXIT1_STATE;
    int exitState2 = INITIAL_EXIT2_STATE;
    int exitState3 = INITIAL_EXIT3_STATE;
    int exitState4 = INITIAL_EXIT4_STATE;
    int exitState5 = INITIAL_EXIT5_STATE;
    int exitState6 = INITIAL_EXIT6_STATE;
    int exitState7 = INITIAL_EXIT7_STATE;

    float timeScaler = INITIAL_TIME_SCALER;

    float steerWeight = INITIAL_STEER_WEIGHT;
    float avoidWeight = INITIAL_AVOID_WEIGHT;
    float collisionWeight = INITIAL_COLLISION_WEIGHT;
    float goalWeight = INITIAL_GOAL_WEIGHT;


    set_TIME_SCALER(&timeScaler);
    set_STEER_WEIGHT(&steerWeight);
    set_AVOID_WEIGHT(&avoidWeight);
    set_COLLISION_WEIGHT(&collisionWeight);
    set_GOAL_WEIGHT(&goalWeight);


    set_EMMISION_RATE_EXIT1(&emmisionRateExit1);
    set_EMMISION_RATE_EXIT2(&emmisionRateExit2);
    set_EMMISION_RATE_EXIT3(&emmisionRateExit3);
    set_EMMISION_RATE_EXIT4(&emmisionRateExit4);
    set_EMMISION_RATE_EXIT5(&emmisionRateExit5);
    set_EMMISION_RATE_EXIT6(&emmisionRateExit6);
    set_EMMISION_RATE_EXIT7(&emmisionRateExit7);

    set_EXIT1_PROBABILITY(&exitProbability1);
    set_EXIT2_PROBABILITY(&exitProbability2);
    set_EXIT3_PROBABILITY(&exitProbability3);
    set_EXIT4_PROBABILITY(&exitProbability4);
    set_EXIT5_PROBABILITY(&exitProbability5);
    set_EXIT6_PROBABILITY(&exitProbability6);
    set_EXIT7_PROBABILITY(&exitProbability7);

    set_EXIT1_STATE(&exitState1);
    set_EXIT2_STATE(&exitState2);
    set_EXIT3_STATE(&exitState3);
    set_EXIT4_STATE(&exitState4);
    set_EXIT5_STATE(&exitState5);
    set_EXIT6_STATE(&exitState6);
    set_EXIT7_STATE(&exitState7);

}
```

**Comment by ptheywood** [comment]
The `<gpu:initFunctions>` element must be *after* the `gpu:functionFiles` element, rather than `<gpu:constants>` which I incorrectly suggested above.

I.e. 

```
<gpu:environment>
        <gpu:constants>
            ...
        </gpu:constants>

        <gpu:functionFiles>
            <file>functions.c</file>
        </gpu:functionFiles>

        <gpu:initFunctions>
            <gpu:initFunction>
                <gpu:name>setInitialConstants</gpu:name>
            </gpu:initFunction>
        </gpu:initFunctions>
</gpu:environment>
```

**Comment by ABallisat** [comment]
Thank you, this fix now works.

**Comment by ABallisat** [comment]
Wasn't sure where to put this but thought here would be a good place to tag it on:

I am trying to set default values for environment variables, such as emission rates, in the XML model file however when I do (by adding a defaultValue tag) these are not being assigned. Looking through the source code (if I have understood it correctly) it does not appear that these values for environment constants are being read and assigned in the readInitialStates function. What code do I need to change to get these to be set? My reason for asking is that I am looking at doing parameter studies with this and don't want to recompile every time I run a model.

**Comment by ptheywood** [comment]
These missing features have both been added in the upcoming 1.5 release, and are available in the `release-1.5` branch (see issue #70) 

The individual commits which added these changes are: 

+ 99f3339 adds support for default values in xmlmodelfiles 
+ 2a5fb7b adds loading of environment values from constants

Although there have been additional changes since then which (may) impact the relevant bits of code.

**Comment by ABallisat** [comment]
Thank you, I have now migrated to release 1.5. I can set the default value in the xml file and when compiled it works fine. If I then change the default value without compiling again (essentially using the XMLModelFile as an input file) it has no effect. Is this the expected behaviour? If so, is there any way that these values can be changed without having to recompile each time they are changed?

**Comment by ptheywood** [comment]
The default values provided in `XMLModelFile.xml` are only used at compile time. I.e. if a value is not provided at runtime in the initial states file. 

The structure of the initial states file for environmental variables is as follows.

```
<states>
	<itno>0</itno>
    <environment>
        <SEED>12</SEED>
        <INIT_POPULATION>3</INIT_POPULATION>
    </environment>
    <xagent>
        ...
    </xagent>
</states>
```

I.e. for this initial states file, the value of the environment constant `SEED` will be set to `12` and `INIT_POPULATION` to `3`. 

I believe I still need to update the FLAME GPU documentation regarding this will whill be completed by the FLAME GPU 1.5 release.

The format of the initial states file differs from the XML Model File quite significantly which might not be particularly clear.

**Comment by ABallisat** [comment]
Yes this does need to be updated as it is significant. Basically in its current form FLAME is not well suited to parameter space mapping as each set of initial conditions will require the model to be recompiled, in my case the compile time is about twice the model run time so this is effectively trebling my model evaluation time. Are there any plans to go to an input file system in the future?

**Comment by ptheywood** [comment]
As of 1.5 you can provide different values per initial states file, ie. for parameter space mapping. 

I.e. Say i have a model with the 2 environmental constants (parameters) from above, called `SEED` and `INIT_POPULATION`. 

In the `XMLModelFile.xml` these are defined within the `<gpu:environment>` tag as follows:
```
    <gpu:constants>
      <gpu:variable>
        <type>unsigned int</type>
        <name>SEED</name>
        <defaultValue>0</defaultValue>
      </gpu:variable>
      <gpu:variable>
        <type>unsigned int</type>
        <name>INIT_POPULATION</name>
        <defaultValue>1</defaultValue>
      </gpu:variable>
    </gpu:constants>
```
This means that if no initial states file is provided, or if the relevant values are not provided in the initial states file, the value of `SEED` will be `0` and the value of `INIT_POPULATION` will be `1`. 

Lets say we wish to run this with `SEED` values `0`,`1`,`2` and `INIT_POPULATION` of `10`,`100`,`1000`, this could be achieved with 9 initial states files (in separate folders to avoid output files overwriting. 

I.e. I have the following file structure
```
iterations
    ├── 0-10
    │   └── 0.xml
    ├── 0-100
    │   └── 0.xml
    ├── 0-1000
    │   └── 0.xml
    ├── 1-10
    │   └── 0.xml
    ├── 1-100
    │   └── 0.xml
    ├── 1-1000
    │   └── 0.xml
    ├── 2-10
    │   └── 0.xml
    ├── 2-100
    │   └── 0.xml
    └── 2-1000
        └── 0.xml
```

The contents of each `0.xml` file would then be different. Assuming my agents are all created dynamically via an init function, each `0.xml` file would look very similar but with different values. 

`0-10/0.xml` would contain:

```
<states>
    <itno>0</itno>
    <environment>
        <SEED>0</SEED>
        <INIT_POPULATION>10</INIT_POPULATION>
    </environment>
</states>
```

`0-100/0.xml` would contain:
```
<states>
    <itno>0</itno>
    <environment>
        <SEED>0</SEED>
        <INIT_POPULATION>100</INIT_POPULATION>
    </environment>
</states>
```

`0-100/0.xml` would contain:
```
<states>
    <itno>0</itno>
    <environment>
        <SEED>0</SEED>
        <INIT_POPULATION>1000</INIT_POPULATION>
    </environment>
</states>
```

`1-10/0.xml` would contain:
```
<states>
    <itno>0</itno>
    <environment>
        <SEED>1</SEED>
        <INIT_POPULATION>10</INIT_POPULATION>
    </environment>
</states>
```


and so on. 

This should work in the current version of FLAME GPU 1.5.

Although for large initial populations (i.e the pedestrian models which contain the environment) this does mean the duplication of (relatively) large XML files for only a very small change.


We are considering how we can automate the task of parameter exploration, although this will not be available until at least version 2 of FLAME GPU (if not later).

**Comment by ABallisat** [comment]
Ah that makes considerably more sense, I think I missed that in the docs. It may be worth including this as an example as I can imagine others will want to do this too. Thanks for your help and for responding so quickly.

**Comment by ptheywood** [comment]
I agree that it is worth documenting this clearly / providing an example. I've opened a few new issues to ensure this is done. 

FLAMEGPU/flamegpu#231 and FLAMEGPU/docs#18

**Comment by mondus** [comment]
Now fixed in 1.5. Docs issues todo. Example could be provided in a separate repo.
